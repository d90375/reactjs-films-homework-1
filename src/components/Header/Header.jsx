import React from 'react';
import Logo from './components/Logo';
import SearchPanel from '../SearchPanel';

import styles from './Header.scss';

const Header = () => (
  <header className={styles.header}>
    <Logo />
    <SearchPanel />
  </header>
);

export default Header;
