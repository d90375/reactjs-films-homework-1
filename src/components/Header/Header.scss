.header {
  position: absolute;
  box-sizing: border-box;
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  max-width: 2400px;
  margin: 0 auto;
  padding: 0 5%;
}
