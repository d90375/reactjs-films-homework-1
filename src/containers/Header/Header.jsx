import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import Logo from '../../components/Logo';
import SearchPanel from '../../components/SearchPanel';

import styles from './Header.scss';

class Header extends Component {
  state = {
    timer: null,
  };

  getSearch = async (event) => {
    let { timer } = this.state;

    if (timer) {
      clearInterval(timer);
    }

    const query = event.target.value;
    const {
      setMoviesCondition,
      fetchMovies,
      setSearchQuery,
      history,
    } = this.props;

    setSearchQuery(query);

    timer = setTimeout(async () => {
      if (query !== '') {
        await setMoviesCondition('Search');
        const { condition } = this.props;
        fetchMovies(condition, query);
        history.push(`/?search=${query}`);
      } else {
        await setMoviesCondition('Trending');
        const { condition } = this.props;
        fetchMovies(condition);
        history.push(`/${condition}`);
      }
    }, 400);
    this.setState({ timer });
  }

  render() {
    const { query } = this.props;
    return (
      <header className={styles.header}>
        <Logo />
        <SearchPanel name="Search" getSearch={this.getSearch} query={query} />
      </header>
    );
  }
}

Header.propTypes = {
  setMoviesCondition: PropTypes.func.isRequired,
  fetchMovies: PropTypes.func.isRequired,
  setSearchQuery: PropTypes.func.isRequired,
  condition: PropTypes.string.isRequired,
  query: PropTypes.string.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

export default withRouter(Header);
