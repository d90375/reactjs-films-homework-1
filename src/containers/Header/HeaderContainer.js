import { connect } from 'react-redux';
import { setMoviesCondition, fetchMovies } from '../../modules/movies/moviesAction';
import { getMoviesCondition } from '../../modules/movies/moviesSelector';
import Header from './Header';

const mapStateToProps = (state) => ({
  condition: getMoviesCondition(state),
});

const mapDispatchToProps = {
  setMoviesCondition,
  fetchMovies,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Header);
