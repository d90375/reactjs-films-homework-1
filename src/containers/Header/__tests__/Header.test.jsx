import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import { create } from 'react-test-renderer';
import Header from '../Header';

describe('Header tests', () => {
  describe('Header render', () => {
    it('Header renders correctly', () => {
      const mockCallback = jest.fn();
      const renderer = new ShallowRenderer();
      renderer.render(<Header
        setMoviesCondition={mockCallback}
        fetchMovies={mockCallback}
        condition="Trending"
      />);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });
  });

  describe('getSearch function', () => {
    it('getSearch function called with query', () => {
      jest.useFakeTimers();

      const mockSetMoviesCondition = jest.fn();
      const mockFetchMovies = jest.fn();

      const tree = create(
        <Router>
          <Header
            setMoviesCondition={mockSetMoviesCondition}
            fetchMovies={mockFetchMovies}
            condition="Trending"
          />
        </Router>,
      );

      const searchPanel = tree.root.findByProps({ name: 'Search' });

      const e = {
        target: {
          value: 'sonic',
        },
      };

      searchPanel.props.getSearch(e);

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(0);

      jest.runAllTimers();

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(1);
    });

    it('getSearch function called without query', () => {
      jest.useFakeTimers();

      const mockSetMoviesCondition = jest.fn();
      const mockFetchMovies = jest.fn();

      const tree = create(
        <Router>
          <Header
            setMoviesCondition={mockSetMoviesCondition}
            fetchMovies={mockFetchMovies}
            condition="Trending"
          />
        </Router>,
      );

      const searchPanel = tree.root.findByProps({ name: 'Search' });

      const e = {
        target: {
          value: '',
        },
      };

      searchPanel.props.getSearch(e);

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(0);

      jest.runAllTimers();

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(1);

      e.target.value = 'sonic';
      searchPanel.props.getSearch(e);

      jest.runAllTimers();

      expect(mockSetMoviesCondition.mock.calls.length).toEqual(2);
    });
  });
});
