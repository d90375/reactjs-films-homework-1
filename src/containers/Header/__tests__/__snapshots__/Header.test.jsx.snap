// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests Header render Header renders correctly 1`] = `
<header
  className="header"
>
  <Logo />
  <SearchPanel
    getSearch={[Function]}
    name="Search"
  />
</header>
`;
